server/cluster/unsafe_recovery_controller.go

```diff
@@ -121,452 +142,629 @@ func (u *unsafeRecoveryController) RemoveFailedStores(failedStores map[uint64]st
-		targetRegions[update.Id] = update
+
+	storeID := heartbeat.Stats.StoreId
+	if _, isFailedStore := u.failedStores[storeID]; isFailedStore {
+		// This should be unreachable.
+		// TODO: abort the process directly
+		return false
 	}
-	for _, del := range u.storeRecoveryPlans[storeID].Deletes {
-		toBeRemovedRegions[del] = true
+
+	if _, find := u.storeReports[storeID]; !find {
+		// No need to collect the report of the store
+		return false
 	}
-	numFinished := 0
-	for _, peerReport := range report.PeerReports {
-		region := peerReport.RegionState.Region
-		if _, ok := toBeRemovedRegions[region.Id]; ok {
-			return false
-		} else if target, ok := targetRegions[region.Id]; ok {
-			if bytes.Equal(target.StartKey, region.StartKey) && bytes.Equal(target.EndKey, region.EndKey) && !u.containsFailedPeers(region) {
-				numFinished += 1
-			} else {
-				return false
+
+	if report, exists := u.storeReports[storeID]; exists && report == nil { // if receive duplicated report from same TiKV, just ignore
+		u.storeReports[storeID] = heartbeat.StoreReport
+		u.numStoresReported++
+		if u.numStoresReported == len(u.storeReports) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (u *unsafeRecoveryController) GetStage() unsafeRecoveryStage {
+	u.Lock()
+	defer u.Unlock()
+	return u.stage
+}
+
+func (u *unsafeRecoveryController) changeStage(stage unsafeRecoveryStage) {
+	u.stage = stage
+
+	var output []string
+	switch u.stage {
+	case idle:
+	case collectReport:
+		output = append(output, "Enter collect report stage")
+	case forceLeader:
+		output = append(output, "Enter force leader stage")
+		output = append(output, u.getForceLeaderPlanDigest()...)
+	case demoteFailedVoter:
+		output = append(output, "Enter demote failed voter stage")
+		output = append(output, u.getDemoteFailedVoterPlanDigest()...)
+	case createEmptyRegion:
+		output = append(output, "Enter create empty region stage")
+		output = append(output, u.getCreateEmptyRegionPlanDigest()...)
+	case finished:
+		output = append(output, "Unsafe recovery finished")
+	}
+
+	u.output = append(u.output, output...)
+	for _, o := range output {
+		log.Info(o)
+	}
+
+	// reset store reports to nil instead of delete, because it relays on the item
+	// to decide which store it needs to collect the report from.
+	for k := range u.storeReports {
+		u.storeReports[k] = nil
+	}
+	u.numStoresReported = 0
+}
+
+func (u *unsafeRecoveryController) getForceLeaderPlanDigest() []string {
+	var output []string
+	for storeID, plan := range u.storeRecoveryPlans {
+		forceLeaders := plan.GetForceLeader()
+		if forceLeaders == nil {
+			continue
+		}
+		regions := ""
+		for i, regionID := range forceLeaders.GetEnterForceLeaders() {
+			regions += fmt.Sprintf("%d", regionID)
+			if i != len(forceLeaders.GetEnterForceLeaders())-1 {
+				regions += ", "
 			}
 		}
+		output = append(output, fmt.Sprintf(" - store %d", storeID))
+		output = append(output, fmt.Sprintf("   - force leader on regions: %s", regions))
```

Comment:
`-` can only work well if there are no other logs.

Reply:
do you have any better idea?

Reply:
I didn't find any way to stringify the zap fields

Reply:
Now it's printed in json structure. And all actions in one stage will be output in one log.